pintc/src/types.rs

```diff
@@ -175,6 +175,56 @@ impl Type {
         })
     }
 
+    pub fn replace_nested_enum_with_int(&mut self) {
+        match self {
+            Type::Array { ty, .. } => ty.replace_nested_enum_with_int(),
+
+            Type::Tuple { fields, .. } => {
+                for field in fields {
+                    field.1.replace_nested_enum_with_int();
+                }
+            }
+
+            Type::Custom { span, .. } => {
+                *self = Type::Primitive {
+                    kind: PrimitiveKind::Int,
+                    span: span.clone(),
+                }
+            }
+
+            Type::Alias { ty, .. } => ty.replace_nested_enum_with_int(),
+
+            Type::Map { ty_from, ty_to, .. } => {
+                ty_from.replace_nested_enum_with_int();
+                ty_to.replace_nested_enum_with_int();
+            }
+
+            Type::Vector { ty, .. } => ty.replace_nested_enum_with_int(),
+
+            Type::Error(_) | Type::Unknown(_) | Type::Any(_) | Type::Primitive { .. } => {}
+        }
+    }
+
+    pub fn has_nested_enum(&self) -> bool {
+        match self {
+            Type::Array { ty, .. } => ty.has_nested_enum(),
+
+            Type::Tuple { fields, .. } => fields.iter().any(|field| field.1.has_nested_enum()),
+
+            Type::Custom { .. } => true,
+
+            Type::Alias { ty, .. } => ty.has_nested_enum(),
+
+            Type::Map { ty_from, ty_to, .. } => {
+                ty_from.has_nested_enum() || ty_to.has_nested_enum()
+            }
+
+            Type::Vector { ty, .. } => ty.has_nested_enum(),
+
+            Type::Error(_) | Type::Unknown(_) | Type::Any(_) | Type::Primitive { .. } => false,
+        }
+    }
+
```

Comment:
Here I've used the same logic to either
1. replace a nested `enum ty` with the `int type` -- for use with `exprs` and `vars`
2. report whether or not there is a nested `enum ty` -- for reporting an error when `enums` are present in `states`

Should these be combined or remain separate?